### ISSUE TITLE
Terminate kernel before sending BinaryFileResponse content during test

### ISSUE THREAD

**Issue by fsevestre** [symfony/symfony#16293]
Hi,

I'm having an issue during testing an action which return a BinaryFileResponse.

My use case is:
- We copy a file from a remote server in a temporary directory on our app ;
- We do stuff on it ;
- We return a BinaryFileResponse ;
- And on the kernel.terminate event we remove the file.

It's working fine but when I try to test it, I have this error:

```
fopen(... /tmp/7fe716dd333c2dc8eec1b0d0f212037608f4de5c.pdf): failed to open stream: No such file or directory

... /vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/BinaryFileResponse.php:266
... /vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Client.php:221
... /vendor/symfony/symfony/src/Symfony/Component/BrowserKit/Client.php:300
... /src/AppBundle/Tests/Tests/Controller/DocumentControllerTest.php:15
... /vendor/phpunit/phpunit/phpunit:47
```

By searching, it seems that the `BinaryFileResponse::sendContent` is called after the terminate kernel event, which is not the case in not testing environment.

Thanks

**Comment by fsevestre** [comment]
I forked the symfony-standard: https://github.com/fsevestre/symfony-standard/tree/16293-issue-terminate-kernel-binary-file-response

**Comment by stof** [comment]
hmm, the Client is indeed sending the response content only after the terminate event, which won't work fine for StreamedResponse and BinaryFileResponse

**Comment by robinduval** [comment]
@stof We understand the point. In our development (production) environment the content is sent before. We don't understand why in test the order is changed.

**Comment by Nicofuma** [comment]
Couldn't you use `$response->deleteFileAfterSend(true);` instead of deleting the file manually in `kernel.terminate`?

**Comment by jvasseur** [comment]
@Nicofuma I think deleteFileAfterSend doesn't work if you use the X-SendFile or X-Accel-Redirect features.

**Comment by fsevestre** [comment]
Thanks @Nicofuma, `$response->deleteFileAfterSend(true);` worked for my use case.

But there is still an _order_ issue here.

**Comment by fnagel** [comment]
Same issue here. Tried using `deleteFileAfterSend` but same effect. Works in non test env but fails due to already deleted file in tests. Any other idea how to workaround this?

**Comment by stof** [comment]
if you use X-Send-File, there is no good way to delete the file automatically: PHP will ends the processing before the file is sent, including the terminate event, as it delegates the actual sending to Nginx.
So deleting the file would have to be done differently, only after Nginx is done with the file. And this would require a feature at the Nginx level (either to remove the file there, or to call PHP again at that point)

**Comment by fnagel** [comment]
Thanks a lot for helping out here! 

Not using X-Send-File but my tests fail when using this code in controller:

````
$response = $app->sendFile($someFile->getRealPath());
$response->deleteFileAfterSend(true);
return $response;
````
and 
````
ob_start();
$client->request('POST', $url);
$client->getResponse()->sendContent();
$content = ob_get_clean();
````
in the tests. 

Doing something similar but using curl works just as expected, even within the same web test class.

**Comment by podorozhny** [comment]
@fnagel +1

@fabpot can u review it please?

**Comment by patrickjahns** [comment]
Also just run into the issue. 

Use-case for me was the point, that I have a simple app that accepts files, modifies the content and sends the file back. I actually wanted to avoid having a cleanup job and thus wrote a Helper that would remove the uploaded contents when `kernel.terminate` event was executed in symfony.

This however broke my test suites due to the above reason. 

Jumped a bit through the code and found:
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/Client.php#L66-L74

Which is called before the filter part ( which actually clears out the buffer )
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/BrowserKit/Client.php#L402-L408

**Comment by carsonbot** [comment]
Hey, thanks for your report!
There has not been a lot of activity here for a while. Is this bug still relevant? Have you managed to find a workaround?

**Comment by carsonbot** [comment]
Hello? This issue is about to be closed if nobody replies.

**Comment by rodnaph** [comment]
I've just run into this limitation (4.4), I think it's still valid @carsonbot

I've worked around the issue in my test, where I know it should be a `BinaryFileResponse`...

```php
$client->request('GET', '/some/url.pdf');
assert($client->getResponse() instanceof BinaryFileResponse);

$browserResponse = $client->getInternalResponse();

$response = new Response(
    $browserResponse->getContent(),
    $browserResponse->getStatusCode(),
    $browserResponse->getHeaders()
);

// use response as normal
```

**Comment by carsonbot** [comment]
Hey, thanks for your report!
There has not been a lot of activity here for a while. Is this bug still relevant? Have you managed to find a workaround?

**Comment by carsonbot** [comment]
Hello? This issue is about to be closed if nobody replies.

**Comment by carsonbot** [comment]
Hey,

I didn't hear anything so I'm going to close it. Feel free to comment if this is still relevant, I can always reopen!

**Comment by TechnoCraftDevops** [comment]
Hi every one, this is my use case (quasi same as @fsevestre )

1. i create two files (text and pdf) in two different directories (`public/fromTemplate/mypdf.pdf` and `public/data/mytest.txt`)
2. i get them in a service to zip them
3. send the zip file in BinaryFileResponse
4. _**here i want to remove them after the sanding of BinaryFileResponse (that's the problem) !**_

The solution can be like this :
```
namespace App\Service;

class FilesToRemoveService
{
    private $filesToRemove = [];

    public function setFilesToRemoveAfterReponse(string $filePath): array
    {
        $this->filesToRemove [] = $filePath;

        return  $this->filesToRemove;
    }

    public function GetFilesToRemoveAfterReponse(): array
    {
        return  $this->filesToRemove;
    }
```
i use this service to store my files :
```
    #[Route('/downloadFile/{id}', name: 'downloadZip')]
    public function downloadZip($id, FileService $fileService, FilesToRemoveService $fileToRemoveService): BinaryFileResponse {
        try {
           // my code that generate files

            $zipFile = $fileService->createZipFile($data); // return pathFile string
            $fileToRemoveService->setFilesToRemoveAfterReponse($zipFile); // I use this method before too so I have a well-filled list

            return new BinaryFileResponse($zipFile, 200, [
                'Content-Type' => 'application/zip',
                'Content-Disposition' => 'inline; filename="' . $zipFile . '"',
            ]);
        } catch (Exception $e) {
            return new JsonResponse(['state' => $e->getMessage(),], $e->getCode());
        }
    }
```
and in my event subscriber link to `onKernelTerminate`  event:
```
<?php

namespace App\EventSubscriber;

use App\Service\FileToRemoveService;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpKernel\Event\TerminateEvent;
use Symfony\Component\HttpKernel\KernelEvents;

class DownloadFileSubscriber implements EventSubscriberInterface
{
    private $fileToRemoveService;

    public function __construct(FileToRemoveService $fileToRemoveService)
    {
        $this->fileToRemoveService = $fileToRemoveService;
    }
  
    public function onKernelTerminate(TerminateEvent $event): void
    {
        $filesToRemove = $this->fileToRemoveService->GetFilesToRemoveAfterReponse();
        foreach ($filesToRemove as $fileToRemove) {
            unlink($fileToRemove);
        }
    }

    public static function getSubscribedEvents(): array
    {
        return [
            KernelEvents::TERMINATE => 'onKernelTerminate',
        ];
    }
}
```
Hope this solution help you.

**Comment by nicolas-grekas** [comment]
Fixed by #65496, merged into the 6.4 branch.
